### ISSUE TITLE
Baseline image reuse breaks parallel testing

### ISSUE THREAD

**Issue by QuLogic** [matplotlib/matplotlib#7673]
Travis has been failing a lot lately, and I'm pretty sure the reason is baseline image reuse + parallel tests. I believe pytest parallelizes across extension and baseline image, so when a test writes to the same baseline more than once, there's a high possibility that one figure will be saved on top of another while it's trying to verify that it was correct.

For example, the `step_linestyle` image has been failing a lot recently, which you can try out with:
```
$ py.test test_axes.py -k linestyle -n 8
...
...
________________________________________________ test_step_linestyle[1-step_linestyle-svg] _________________________________________________
[gw7] linux -- Python 3.5.2 /home/elliott/code/conda/envs/mpl35/bin/python
expected = '/home/elliott/code/matplotlib/lib/matplotlib/tests/result_images/test_axes/step_linestyle-expected_svg.png'
actual = '/home/elliott/code/matplotlib/lib/matplotlib/tests/result_images/test_axes/step_linestyle_svg.png'
tol = 0, in_decorator = True

    def compare_images(expected, actual, tol, in_decorator=False):
        """
        Compare two "image" files checking differences within a tolerance.
    
        The two given filenames may point to files which are convertible to
        PNG via the `.converter` dictionary. The underlying RMS is calculated
        with the `.calculate_rms` function.
    
        Parameters
        ----------
        expected : str
            The filename of the expected image.
        actual :str
            The filename of the actual image.
        tol : float
            The tolerance (a color value difference, where 255 is the
            maximal difference).  The test fails if the average pixel
            difference is greater than this value.
        in_decorator : bool
            If called from image_comparison decorator, this should be
            True. (default=False)
    
        Example
        -------
        img1 = "./baseline/plot.png"
        img2 = "./output/plot.png"
        compare_images( img1, img2, 0.001 ):
    
        """
        if not os.path.exists(actual):
            msg = "Output image %s does not exist." % actual
            raise Exception(msg)
    
        if os.stat(actual).st_size == 0:
            msg = "Output image file %s is empty." % actual
            raise Exception(msg)
    
        verify(actual)
    
        # Convert the image to png
        extension = expected.split('.')[-1]
    
        if not os.path.exists(expected):
            raise IOError('Baseline image %r does not exist.' % expected)
    
        if extension != 'png':
            actual = convert(actual, False)
            expected = convert(expected, True)
    
        # open the image files and remove the alpha channel (if it exists)
        expectedImage = _png.read_png_int(expected)
>       actualImage = _png.read_png_int(actual)
E       SystemError: <built-in function read_png_int> returned NULL without setting an error
```
which failed reading the png-from-svg, but running again gets:
```
________________________________________________ test_step_linestyle[0-step_linestyle-svg] _________________________________________________
[gw3] linux -- Python 3.5.2 /home/elliott/code/conda/envs/mpl35/bin/python
expected = '/home/elliott/code/matplotlib/lib/matplotlib/tests/result_images/test_axes/step_linestyle-expected.svg'
actual = '/home/elliott/code/matplotlib/lib/matplotlib/tests/result_images/test_axes/step_linestyle.svg'
tol = 0, in_decorator = True

    def compare_images(expected, actual, tol, in_decorator=False):
        """
        Compare two "image" files checking differences within a tolerance.
    
        The two given filenames may point to files which are convertible to
        PNG via the `.converter` dictionary. The underlying RMS is calculated
        with the `.calculate_rms` function.
    
        Parameters
        ----------
        expected : str
            The filename of the expected image.
        actual :str
            The filename of the actual image.
        tol : float
            The tolerance (a color value difference, where 255 is the
            maximal difference).  The test fails if the average pixel
            difference is greater than this value.
        in_decorator : bool
            If called from image_comparison decorator, this should be
            True. (default=False)
    
        Example
        -------
        img1 = "./baseline/plot.png"
        img2 = "./output/plot.png"
        compare_images( img1, img2, 0.001 ):
    
        """
        if not os.path.exists(actual):
            msg = "Output image %s does not exist." % actual
            raise Exception(msg)
    
        if os.stat(actual).st_size == 0:
            msg = "Output image file %s is empty." % actual
>           raise Exception(msg)
E           Exception: Output image file /home/elliott/code/matplotlib/lib/matplotlib/tests/result_images/test_axes/step_linestyle.svg is empty.
```
which points to a truncated svg.

To fix this, we should be able to stop pytest from parallelizing across baseline images, but what if more than one test uses the same image? I haven't yet checked that case, but it's a possibility. Making some kind of lock-per-baseline-image seems like too much work. Should we start duplicating our baseline images if that's happening?

**Comment by QuLogic** [comment]
Here is a list of tests that reuse the same baseline image on themselves:
```
$ grep -IR -A2 'baseline_images=\[\(\(['"'"'"]\)[A-Za-z0-9_-]\+\2\), \1'
test_cycles.py:@image_comparison(baseline_images=['marker_cycle', 'marker_cycle'],
test_cycles.py-                  remove_text=True, extensions=['png'])
test_cycles.py-def test_marker_cycle():
--
test_axes.py:@image_comparison(baseline_images=['single_point', 'single_point'])
test_axes.py-def test_single_point():
test_axes.py-    # Issue #1796: don't let lines.marker affect the grid
--
test_axes.py:@image_comparison(baseline_images=['hexbin_extent', 'hexbin_extent'],
test_axes.py-                  remove_text=True, extensions=['png'])
test_axes.py-def test_hexbin_extent():
--
test_axes.py:@image_comparison(baseline_images=['imshow', 'imshow'],
test_axes.py-                  remove_text=True)
test_axes.py-def test_imshow():
--
test_axes.py:@image_comparison(baseline_images=['hist2d', 'hist2d'])
test_axes.py-def test_hist2d():
test_axes.py-    np.random.seed(0)
--
test_axes.py:@image_comparison(baseline_images=['scatter', 'scatter'])
test_axes.py-def test_scatter_plot():
test_axes.py-    fig, ax = plt.subplots()
--
test_axes.py:@image_comparison(baseline_images=['boxplot', 'boxplot'],
test_axes.py-                  tol=1,
test_axes.py-                  style='default')
--
test_axes.py:@image_comparison(baseline_images=['eventplot', 'eventplot'], remove_text=True)
test_axes.py-def test_eventplot():
test_axes.py-    '''
--
test_axes.py:@image_comparison(baseline_images=['step_linestyle', 'step_linestyle'],
test_axes.py-                  remove_text=True)
test_axes.py-def test_step_linestyle():
--
test_axes.py:@image_comparison(baseline_images=['twin_spines_on_top', 'twin_spines_on_top'],
test_axes.py-                  extensions=['png'], remove_text=True)
test_axes.py-def test_twin_spines_on_top():
--
test_axes.py:@image_comparison(baseline_images=['pie_linewidth_0', 'pie_linewidth_0',
test_axes.py-                                   'pie_linewidth_0'],
test_axes.py-                  extensions=['png'])
```

**Comment by QuLogic** [comment]
There are a couple of baseline images that are reused in separate tests:
```
$ grep -IRo 'baseline_images=\[\(['"'"'"]\)\([A-Za-z0-9_-]\)\+\1' | cut -d[ -f2 | sort | uniq -c | grep -v ' 1 '
      2 'imshow'
      2 'log_scales'

$ grep -IRA2 \'imshow\'
test_axes.py:@image_comparison(baseline_images=['imshow', 'imshow'],
test_axes.py-                  remove_text=True)
test_axes.py-def test_imshow():
--
test_image.py:@image_comparison(baseline_images=['imshow'], remove_text=True)
test_image.py-def test_imshow():
test_image.py-    import numpy as np
$ grep -IRA2 \'log_scales\'
test_axes.py:@image_comparison(baseline_images=['log_scales'])
test_axes.py-def test_log_scales():
test_axes.py-    fig = plt.figure()
--
test_scale.py:@image_comparison(baseline_images=['log_scales'], remove_text=True)
test_scale.py-def test_log_scales():
test_scale.py-    ax = plt.subplot(122, yscale='log', xscale='symlog')
```

**Comment by Kojoley** [comment]
Closed with #7674
